datashader/core.py

```diff
@@ -317,16 +318,41 @@ def raster(self,
         if interpolate not in upsample_methods.keys():
             raise ValueError('Invalid upsample method: options include {}'.format(list(upsample_methods.keys())))
 
-        if hasattr(agg,"column"):
-            # Accepts a reduction from ./reductions.py, but only uses the class name and the optional column name
-            if hasattr(source,"name") and agg.column != source.name:
-                raise ValueError('data variable {0} not found in array with name {1}'.format(agg.column,source.name))
-            agg=type(agg)
-        
+        if not isinstance(source, (DataArray, Dataset)):
+            raise ValueError('Expected xarray DataArray or Dataset as '
+                             'the data source, found %s.'
+                             % type(source).__name__)
+
+        column = None
+        if isinstance(agg, rd.Reduction):
+            agg, column = type(agg), agg.column
+            if (isinstance(source, DataArray) and column is not None
+                and source.name != column):
+                agg_repr = '%s(%r)' % (agg.__name__, column)
+                raise ValueError('DataArray name %r does not match '
+                                 'supplied reduction %s.' %
+                                 (source.name, agg_repr))
+
+        if isinstance(source, Dataset):
+            data_vars = list(source.data_vars)
+            if column is None:
+                raise ValueError('When supplying a Dataset the agg reduction '
+                                 'must specify the variable to aggregate. '
+                                 'Available data_vars include: %r.' % data_vars)
+            elif column not in source.data_vars:
+                raise KeyError('Supplied reduction column %r not found '
+                               'in Dataset, expected one of the following '
+                               'data variables: %r.' % (column, data_vars))
+            source = source[column]
+
         if agg not in downsample_methods.keys():
             raise ValueError('Invalid downsample method: options include {}'.format(list(downsample_methods.keys())))
         ds_method = downsample_methods[agg]
 
+        if source.ndim not in [2, 3]:
+            raise ValueError('Raster aggregation expects a 2D or 3D '
+                             'DataArray, found %s dimensions' % source.ndim)
```

Comment:
Should this say "expects a 2D DataArray or a 3D Dataset"?  How would we handle a 3D DataArray?

Reply:
No, this is correct. The operation handles arrays with multiple layers as long as the first dimension defines the layer, e.g. to resample all the layers in the landsat data at once or a time dimension. It's specifically to handle whatever``xr.open_rasterio`` but can be used in other cases too.

Reply:
Ok.  It would be nice if the message mentioned Dataset to make it clear that's an acceptable way to pass the DataArray, but I can't think of any way to word that that makes it less rather than more confusing, so I guess it's fine as is.